File: django-python-backend/apps/api/errors/http_errors.py
from apps.api.errors.app_error import AppError


class ValidationError(AppError):
    def __init__(self, detail: str = 'Validation failed') -> None:
        super().__init__(
            400,
            'VALIDATION_ERROR',
            'Bad Request',
            detail,
            'https://api.example.dev/errors/validation',
        )


class AuthenticationError(AppError):
    def __init__(self, detail: str = 'Authentication required') -> None:
        super().__init__(
            401,
            'AUTHENTICATION_ERROR',
            'Unauthorized',
            detail,
            'https://api.example.dev/errors/authentication',
        )


class AuthorizationError(AppError):
    def __init__(self, detail: str = 'Forbidden') -> None:
        super().__init__(
            403,
            'AUTHORIZATION_ERROR',
            'Forbidden',
            detail,
            'https://api.example.dev/errors/authorization',
        )


class NotFoundError(AppError):
    def __init__(self, detail: str = 'Resource not found') -> None:
        super().__init__(
            404, 'NOT_FOUND', 'Not Found', detail, 'https://api.example.dev/errors/not-found'
        )


class ConflictError(AppError):
    def __init__(self, detail: str = 'Conflict') -> None:
        super().__init__(
            409, 'CONFLICT', 'Conflict', detail, 'https://api.example.dev/errors/conflict'
        )


class RateLimitExceededError(AppError):
    def __init__(self, detail: str = 'Rate limit exceeded') -> None:
        super().__init__(
            429,
            'RATE_LIMIT_EXCEEDED',
            'Too Many Requests',
            detail,
            'https://api.example.dev/errors/rate-limit-exceeded',
        )


class ExternalServiceError(AppError):
    def __init__(self, detail: str = 'External service failed') -> None:
        super().__init__(
            502,
            'EXTERNAL_SERVICE_ERROR',
            'Bad Gateway',
            detail,
            'https://api.example.dev/errors/external-service',
        )


class ServiceUnavailableError(AppError):
    def __init__(self, detail: str = 'Service unavailable') -> None:
        super().__init__(
            503,
            'SERVICE_UNAVAILABLE',
            'Service Unavailable',
            detail,
            'https://api.example.dev/errors/service-unavailable',
        )


class InternalServerError(AppError):
    def __init__(self, detail: str = 'Internal server error') -> None:
        super().__init__(
            500,
            'INTERNAL_SERVER_ERROR',
            'Internal Server Error',
            detail,
            'https://api.example.dev/errors/internal',
            is_operational=False,
        )

File: springboot-java-backend/src/main/java/com/example/template/errors/HttpErrors.java
package com.example.template.errors;

public final class HttpErrors {
  private HttpErrors() {}

  public static class ValidationError extends AppError {
    public ValidationError(String message) {
      super("VALIDATION_ERROR", 400, "https://example.com/errors/validation", message);
    }
  }

  public static class AuthenticationError extends AppError {
    public AuthenticationError(String message) {
      super("AUTHENTICATION_ERROR", 401, "https://example.com/errors/authentication", message);
    }
  }

  public static class AuthorizationError extends AppError {
    public AuthorizationError(String message) {
      super("AUTHORIZATION_ERROR", 403, "https://example.com/errors/authorization", message);
    }
  }

  public static class NotFoundError extends AppError {
    public NotFoundError(String message) {
      super("NOT_FOUND", 404, "https://example.com/errors/not-found", message);
    }
  }

  public static class ConflictError extends AppError {
    public ConflictError(String message) {
      super("CONFLICT", 409, "https://example.com/errors/conflict", message);
    }
  }

  public static class RateLimitExceededError extends AppError {
    public RateLimitExceededError(String message) {
      super("RATE_LIMIT_EXCEEDED", 429, "https://example.com/errors/rate-limit", message);
    }
  }

  public static class ExternalServiceError extends AppError {
    public ExternalServiceError(String message) {
      super("EXTERNAL_SERVICE_ERROR", 502, "https://example.com/errors/external-service", message);
    }
  }

  public static class ServiceUnavailableError extends AppError {
    public ServiceUnavailableError(String message) {
      super("SERVICE_UNAVAILABLE", 503, "https://example.com/errors/service-unavailable", message);
    }
  }

  public static class InternalServerError extends AppError {
    public InternalServerError(String message) {
      super("INTERNAL_SERVER_ERROR", 500, "https://example.com/errors/internal", message);
    }
  }
}
